### PR TITLE
feat: Add projection to FilterExec

### LINKED ISSUE #5436: Add projection to `FilterExec` to avoid unecessary output creation
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
Currently `FilterExec` filters/outputs all columns present in the filter expression. This does unnecessary work, as we might not need certain columns later.

**Describe the solution you'd like**
Add projection to Filter/FilterExec and use it before passing the `RecordBatch` to the arrow kernel.
This saves a bit of computation, especially when there is only one column in the projection (fast case).

Also update the `schema` based on the projection.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by eejbyfeldt**
## Which issue does this PR close?

Closes #5436.

## Rationale for this change

Improves performance by avoiding copying some of the columns during execution of FilterExec.

Comparision tpch_mem10

```
┏━━━━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Query        ┃      main ┃ filter-projection ┃        Change ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ QQuery 1     │ 2177.83ms │         2127.02ms │     no change │
│ QQuery 2     │  482.20ms │          471.44ms │     no change │
│ QQuery 3     │ 1118.91ms │         1093.83ms │     no change │
│ QQuery 4     │  799.31ms │          628.54ms │ +1.27x faster │
│ QQuery 5     │ 1841.60ms │         1870.31ms │     no change │
│ QQuery 6     │  118.43ms │          117.54ms │     no change │
│ QQuery 7     │ 3323.14ms │         3308.58ms │     no change │
│ QQuery 8     │ 1451.20ms │         1447.37ms │     no change │
│ QQuery 9     │ 2983.42ms │         2981.28ms │     no change │
│ QQuery 10    │ 1434.72ms │         1409.87ms │     no change │
│ QQuery 11    │  950.67ms │          956.36ms │     no change │
│ QQuery 12    │  339.26ms │          322.72ms │     no change │
│ QQuery 13    │  947.67ms │          761.68ms │ +1.24x faster │
│ QQuery 14    │  163.97ms │          171.94ms │     no change │
│ QQuery 15    │  384.72ms │          402.21ms │     no change │
│ QQuery 16    │  371.63ms │          373.32ms │     no change │
│ QQuery 17    │ 3024.34ms │         3098.19ms │     no change │
│ QQuery 18    │ 7221.31ms │         7250.94ms │     no change │
│ QQuery 19    │  266.89ms │          262.75ms │     no change │
│ QQuery 20    │  815.55ms │          824.50ms │     no change │
│ QQuery 21    │ 4398.76ms │         4243.96ms │     no change │
│ QQuery 22    │  182.82ms │          180.98ms │     no change │
└──────────────┴───────────┴───────────────────┴───────────────┘
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┓
┃ Benchmark Summary                ┃            ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━┩
│ Total Time (main)                │ 34798.35ms │
│ Total Time (filter-projection)   │ 34305.29ms │
│ Average Time (main)              │  1581.74ms │
│ Average Time (filter-projection) │  1559.33ms │
│ Queries Faster                   │          2 │
│ Queries Slower                   │          0 │
│ Queries with No Change           │         20 │
└──────────────────────────────────┴────────────┘

```

## What changes are included in this PR?

Adds a projection to FilterExec similarly to https://github.com/apache/datafusion/pull/9236 for HashJoinExec. It also resuses a lot of the logic from that PR to implement the pushdown.

## Are these changes tested?
Existing tests.

## Are there any user-facing changes?
Yes, projections can now be pushed into the FilterExec.

### CODE REVIEW

File: datafusion/sqllogictest/test_files/aggregate.slt [hunk 1]
```diff
@@ -5772,15 +5772,14 @@ logical_plan
...
-01)ProjectionExec: expr=[v1@0 as v1, v2@1 as v2]
-02)--CoalesceBatchesExec: target_batch_size=8192
-03)----FilterExec: max(having_test.v1)@2 = 3
-04)------AggregateExec: mode=FinalPartitioned, gby=[v1@0 as v1, v2@1 as v2], aggr=[max(having_test.v1)]
-05)--------CoalesceBatchesExec: target_batch_size=8192
-06)----------RepartitionExec: partitioning=Hash([v1@0, v2@1], 4), input_partitions=4
-07)------------RepartitionExec: partitioning=RoundRobinBatch(4), input_partitions=1
-08)--------------AggregateExec: mode=Partial, gby=[v1@0 as v1, v2@1 as v2], aggr=[max(having_test.v1)]
-09)----------------MemoryExec: partitions=1, partition_sizes=[1]
+01)CoalesceBatchesExec: target_batch_size=8192
```
Comment (Dandandan) [inline datafusion/sqllogictest/test_files/aggregate.slt:5775]: I would expect that tpch sqllogictests would also be changed?

Comment (eejbyfeldt) [reply datafusion/sqllogictest/test_files/aggregate.slt:5775]: Updated them now, was not aware of them until now :)

**Review by Dandandan — APPROVED** [review]
Looks very nice 🚀

**Review by berkaysynnada — COMMENTED** [review]
LGTM, thank you @eejbyfeldt. It seems like the `with_projection()` API could end up being a method of the `ExecutionPlan` in the long run, wdyt? (It would improve the performance of all operators which internally refer to some columns but its downstream operators do not actually require that column)

**Comment by eejbyfeldt** [comment]
> It seems like the `with_projection()` API could end up being a method of the `ExecutionPlan` in the long run, wdyt? (It would improve the performance of all operators which internally refer to some columns but its downstream operators do not actually require that column)

Probably worth considering in the future. The argument against making it part of ExecutionPlan is that there are some operators like `CoalescePartitionsExec` where it will not make sense.


Addressed the comments.

File: datafusion/physical-expr/src/equivalence/projection.rs [hunk 2]
```diff
@@ -82,6 +82,11 @@ impl ProjectionMapping {
             .map(|map| Self { map })
     }
 
+    pub fn from_indices(indices: &[usize], schema: &SchemaRef) -> Result<Self> {
```
[the remark this answers is not in the document]
Comment (eejbyfeldt) [reply datafusion/physical-expr/src/equivalence/projection.rs:89]: Fixed in https://github.com/apache/datafusion/pull/12281/commits/0c05fb904f9d2a70f9c656a105f194584189c017

File: datafusion/core/src/physical_optimizer/projection_pushdown.rs [hunk 3]
```diff
@@ -535,11 +538,27 @@ fn try_pushdown_through_union(
     Ok(Some(Arc::new(UnionExec::new(new_children))))
 }
 
+trait EmbededProjection: ExecutionPlan + Sized {
```
[the remark this answers is not in the document]
Comment (eejbyfeldt) [reply datafusion/core/src/physical_optimizer/projection_pushdown.rs]: Fixed in https://github.com/apache/datafusion/pull/12281/commits/0c05fb904f9d2a70f9c656a105f194584189c017

File: datafusion/sqllogictest/test_files/aggregate.slt [hunk 4]
```diff
@@ -5772,15 +5772,14 @@ logical_plan
...
+02)--FilterExec: max(having_test.v1)@2 = 3, projection=[v1@0, v2@1]
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/aggregate.slt:5776]: this is very cool

File: datafusion/sqllogictest/test_files/tpch/q2.slt.part [hunk 5]
```diff
@@ -123,64 +123,61 @@ physical_plan
 22)------------------------------------------HashJoinExec: mode=Partitioned, join_type=Inner, on=[(p_partkey@0, ps_partkey@0)], projection=[p_partkey@0, p_mfgr@1, ps_suppkey@3, ps_supplycost@4]
 23)--------------------------------------------CoalesceBatchesExec: target_batch_size=8192
 24)----------------------------------------------RepartitionExec: partitioning=Hash([p_partkey@0], 4), input_partitions=4
-25)------------------------------------------------ProjectionExec: expr=[p_partkey@0 as p_partkey, p_mfgr@1 as p_mfgr]
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/tpch/q2.slt.part:126]: these plans are doubly good -- they save two copies -- one in `FilterExec` and one in `CoalesceBatchesExec`

**Review by alamb — APPROVED** [review]
This is a really nice improvement -- thank you @eejbyfeldt .

Thank you @Dandandan and @berkaysynnada for the reviews

I went throught he plan changes carefully and they are 👌 👨‍🍳

**Comment by alamb** [comment]
🚀

**Comment by alamb** [comment]
FWIW I ran clickbench and also saw improvement on Q30 and Q31

```
--------------------
Benchmark clickbench_1.json
--------------------
┏━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Query        ┃  main_base ┃ implement-filter-exec-projection ┃        Change ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ QQuery 0     │     0.66ms │                           0.67ms │     no change │
│ QQuery 1     │    70.03ms │                          73.91ms │  1.06x slower │
│ QQuery 2     │   126.45ms │                         124.52ms │     no change │
│ QQuery 3     │   129.91ms │                         137.13ms │  1.06x slower │
│ QQuery 4     │   960.79ms │                         939.71ms │     no change │
│ QQuery 5     │  1036.28ms │                        1022.72ms │     no change │
│ QQuery 6     │    66.12ms │                          64.91ms │     no change │
│ QQuery 7     │    73.34ms │                          73.79ms │     no change │
│ QQuery 8     │  1344.30ms │                        1334.81ms │     no change │
│ QQuery 9     │  1306.69ms │                        1309.81ms │     no change │
│ QQuery 10    │   446.44ms │                         442.18ms │     no change │
│ QQuery 11    │   478.30ms │                         481.05ms │     no change │
│ QQuery 12    │  1113.12ms │                        1104.71ms │     no change │
│ QQuery 13    │  2101.60ms │                        2051.31ms │     no change │
│ QQuery 14    │  1506.83ms │                        1503.34ms │     no change │
│ QQuery 15    │  1064.88ms │                        1060.89ms │     no change │
│ QQuery 16    │  2759.74ms │                        2710.97ms │     no change │
│ QQuery 17    │  2701.32ms │                        2669.99ms │     no change │
│ QQuery 18    │  5503.13ms │                        5371.89ms │     no change │
│ QQuery 19    │   117.10ms │                         119.67ms │     no change │
│ QQuery 20    │  1607.88ms │                        1640.09ms │     no change │
│ QQuery 21    │  1952.27ms │                        1996.70ms │     no change │
│ QQuery 22    │  4382.22ms │                        4405.65ms │     no change │
│ QQuery 23    │ 10759.26ms │                       10851.10ms │     no change │
│ QQuery 24    │   714.73ms │                         718.81ms │     no change │
│ QQuery 25    │   627.59ms │                         634.67ms │     no change │
│ QQuery 26    │   784.53ms │                         784.22ms │     no change │
│ QQuery 27    │  2465.46ms │                        2514.45ms │     no change │
│ QQuery 28    │ 15630.39ms │                       15474.96ms │     no change │
│ QQuery 29    │   572.17ms │                         551.60ms │     no change │
│ QQuery 30    │  1229.86ms │                        1167.28ms │ +1.05x faster │
│ QQuery 31    │  1277.91ms │                        1212.40ms │ +1.05x faster │
│ QQuery 32    │  4446.44ms │                        4522.52ms │     no change │
│ QQuery 33    │  5049.17ms │                        5044.77ms │     no change │
│ QQuery 34    │  4978.22ms │                        4890.22ms │     no change │
│ QQuery 35    │  1759.29ms │                        1734.22ms │     no change │
│ QQuery 36    │   319.80ms │                         312.65ms │     no change │
│ QQuery 37    │   212.58ms │                         204.92ms │     no change │
│ QQuery 38    │   189.61ms │                         189.14ms │     no change │
│ QQuery 39    │   966.16ms │                         982.35ms │     no change │
│ QQuery 40    │    89.98ms │                          86.07ms │     no change │
│ QQuery 41    │    79.43ms │                          80.69ms │     no change │
│ QQuery 42    │    97.07ms │                          95.90ms │     no change │
└──────────────┴────────────┴──────────────────────────────────┴───────────────┘
```

Which correspond to several queries where the filter is on a string column (`"SearchPhrase"` that is not used after the filter

**Pull request merged**